### PR TITLE
Add provisioning api

### PR DESCRIPTION
**Pull request by patrick-tolosa**
- Implement the Provisioning API.
- Test coverage 100% of required arguments -  (did not implement a test for all optional arguments)
- Created CRUD apis for **USERS**, **SUB_ACCOUNTS** and **USER_GROUPS**
- Happy to add more tests if needed

### CODE REVIEW

File: lib/config.js [hunk 1]
```diff
@@ -52,26 +95,14 @@ module.exports = function (new_config, new_value) {
...
-          cloudinary_config[key] = value;
-        }
-      });
-      if (uri.query != null) {
-        entries(uri.query).forEach(([key, value]) => putNestedValue(cloudinary_config, key, value));
-      }
-    }
+    [CLOUDINARY_ENV_URL, CLOUDINARY_ENV_ACCOUNT_URL].forEach((ENV_URL) => {
+      let parsedConfig = getParsedConfigFromEnv(ENV_URL);
+      enrichCloudinaryConfig(parsedConfig);
```
Comment (patrick-tolosa) [inline lib/config.js]: rename enrichConfigWithURIQuery to match enrichCloudinaryConfig

Comment (patrick-tolosa) [inline lib/config.js]: Change enrich to merge

File: lib/network/call_account_api.js [hunk 2]
```diff
@@ -0,0 +1,134 @@
...
+
+const {extend, includes, ensurePresenceOf} = utils;
+
+function call_account_api(method, uri, params, callback, options) {
+  method = method.toUpperCase();
+  let handle_response, query_params;
+
+  ensurePresenceOf({ method, uri });
+  const deferred = Q.defer();
+  const cloudinary = ensureOption(options, "upload_prefix", "https://api.cloudinary.com");
```
Comment (patrick-tolosa) [inline lib/network/call_account_api.js]: See if we can extract this to share some code with call_api.js

File: lib/config.js [hunk 3]
```diff
@@ -44,6 +44,49 @@ function putNestedValue(params, key, value) {
   return params;
 }
 
+function parseCloudinaryConfigFromEnvURL(ENV_STR) {
```
Comment (strausr) [inline lib/config.js]: Should we use camelcase naming convention for parameters and function names?

File: lib/network/call_account_api.js [hunk 4]
```diff
@@ -0,0 +1,134 @@
...
+const https = /^http:/.test(config().upload_prefix) ? require('http') : require('https');
+const querystring = require("querystring");
+const Q = require('q');
+const url = require('url');
+const utils = require("../utils");
+const ensureOption = require('../utils/ensureOption').defaults(config());
+
+const {extend, includes, ensurePresenceOf} = utils;
+
+function call_account_api(method, uri, params, callback, options) {
```
Comment (strausr) [inline lib/network/call_account_api.js]: Very similar to call_api(). Can we re-use call_api and just add the changes before in call_acount_api? 
Also, should we move callback to the end?

File: lib/network/call_account_api.js (hunk 2 above)
Comment (patrick-tolosa) [reply lib/network/call_account_api.js]: Done

File: lib/config.js (hunk 3 above)
Comment (patrick-tolosa) [reply lib/config.js]: Refactored to underscore all the public APIs, I kept internals as camelCase as there's a mixed style in the code already, might be worth slowly migrating to camelCase?

File: lib/network/call_account_api.js (hunk 4 above)
Comment (patrick-tolosa) [reply lib/network/call_account_api.js]: Refactored to make_request, call_api and call_account_api are now very thin layers over it.

File: lib/config.js [hunk 5]
```diff
@@ -44,6 +44,49 @@ function putNestedValue(params, key, value) {
...
+  let conf = {};
+
+  let uri = url.parse(ENV_STR, true);
+
+  if (uri.protocol === 'cloudinary:') {
+    conf = Object.assign({}, conf, {
+      cloud_name: uri.host,
```
Comment (const-cloudinary) [inline lib/config.js]: I would extract common parsing to a separate function and reuse it in `account:` and `cloudinary:` as follows:

generic function parseURL(url)

returns:

{
    host: ...
    user: ...
    pass: ...
    path: ...(without leading char)
}

And  after that you just map `host` to `cloud_name` or `account_id` accordingly, `user` and `pass` as well, and it can be further reused if we have more types of config URLs.

File: lib/network/call_account_api.js [hunk 6]
```diff
@@ -0,0 +1,22 @@
...
+  ensurePresenceOf({ method, uri });
+  const cloudinary = ensureOption(options, "upload_prefix", "https://api.cloudinary.com");
+  const account_id = ensureOption(options, "account_id");
+  const api_url = [cloudinary, "v1_1", "provisioning", "accounts", account_id].concat(uri).join("/");
+  const auth = {
+    api_key: ensureOption(options, "provisioning_api_key"),
+    api_secret: ensureOption(options, "provisioning_api_secret"),
+  };
+
+  return make_request(method, params, auth, api_url, callback, options);
```
Comment (const-cloudinary) [inline lib/network/call_account_api.js]: `perform` or `execute` request looks more appropriate. `make` sounds more like `prepare` request.

File: lib/network/call_account_api.js [hunk 7]
```diff
@@ -0,0 +1,22 @@
...
+
+const { ensurePresenceOf } = utils;
+
+function call_account_api(method, uri, params, callback, options) {
+  ensurePresenceOf({ method, uri });
+  const cloudinary = ensureOption(options, "upload_prefix", "https://api.cloudinary.com");
+  const account_id = ensureOption(options, "account_id");
+  const api_url = [cloudinary, "v1_1", "provisioning", "accounts", account_id].concat(uri).join("/");
+  const auth = {
+    api_key: ensureOption(options, "provisioning_api_key"),
```
Comment (const-cloudinary) [inline lib/network/call_account_api.js]: This is confusing, why don't you call `api_key` and `api_secret` just `key` and `secret`(or user/pass), to avoid ambiguous interpretation of the `cloudinary:` api_key and api_secret

File: lib/provisioning/account.js [hunk 8]
```diff
@@ -0,0 +1,292 @@
...
+ * @param [callback] {function}
+ */
+function sub_account(sub_account_id, options = {}, callback) {
+  let uri = ['sub_accounts', sub_account_id];
+  return call_account_api('GET', uri, {}, callback, options);
+}
+
+
+/**
+ *
```
Comment (const-cloudinary) [inline lib/provisioning/account.js]: Missing @desc

File: lib/provisioning/account.js [hunk 9]
```diff
@@ -0,0 +1,292 @@
+const call_account_api = require('../network/call_account_api');
+
+/**
+ * @desc - List all sub_accounts under the account.
```
Comment (const-cloudinary) [inline lib/provisioning/account.js]: Description should be in 3rd person, `Lists all sub accounts...` (mind `_` as well)

File: lib/provisioning/account.js [hunk 10]
```diff
@@ -0,0 +1,292 @@
...
+
+
+/**
+ *
+ * @param name {string} The display name as shown in the management console.
+ * @param cloud_name {string} A case-insensitive cloud name comprised of alphanumeric and underscore characters. Use with caution - generates an error if the cloud name is not unique across all Cloudinary accounts. Default: a unique cloud name automatically generated by Cloudinary.
+ * @param custom_attributes {object} Any custom attributes you want to associate with the sub-account, as a map/hash of key/value pairs.
+ * @param enabled {boolean} Whether the sub-account is enabled. Default: true
+ * @param base_account {string} The ID of another sub-account, from which to copy all of the following settings: Size limits, Timed limits, and Flags.
+ * @param options
```
Comment (const-cloudinary) [inline lib/provisioning/account.js]: Options description is not consistent across functions, please align

File: lib/provisioning/account.js [hunk 11]
```diff
@@ -0,0 +1,292 @@
...
+    email,
+    role,
+    sub_account_ids: sub_account_ids,
+  };
+  options.content_type = 'json';
+  return call_account_api('POST', uri, params, callback, options);
+}
+
+/**
+ *
```
Comment (const-cloudinary) [inline lib/provisioning/account.js]: Missing desc

File: lib/network/call_account_api.js [hunk 12]
```diff
@@ -1,134 +1,22 @@
 // eslint-disable-next-line import/order
 const config = require("../config");
-const https = /^http:/.test(config().upload_prefix) ? require('http') : require('https');
-const querystring = require("querystring");
-const Q = require('q');
-const url = require('url');
```
Comment (asisayag2) [inline lib/network/call_account_api.js]: @patrick-tolosa , why were those removed? just clean up that is not related to the PR?

File: lib/network/call_account_api.js (hunk 6 above)
Comment (asisayag2) [reply lib/network/call_account_api.js]: +1

File: lib/network/call_account_api.js (hunk 12 above)
Comment (asisayag2) [reply lib/network/call_account_api.js]: Got it... extracted.

File: lib/provisioning/account.js [hunk 13]
```diff
@@ -47,118 +57,118 @@ function createSubAccount(name, cloudName, customAttributes, enabled, baseAccoun
 }
 
 /**
- * @desc - Delete a subAccount by subAccountID
- * @param subAccountId {string} - The ID generated when using createSubAccount
+ * @desc - Delete the specified sub-account. Supported only for accounts with fewer than 1000 images.
```
Comment (asisayag2) [inline lib/provisioning/account.js]: What happens when I try to delete a larger sub-account?
BTW, it says "supported only for **accounts**..." meaning, if I have an account with 2 sub-accounts - one with >1000 images and one with 3 images, I still can't delete the small sub-account?

File: lib/provisioning/account.js [hunk 14]
```diff
@@ -47,118 +57,118 @@ function createSubAccount(name, cloudName, customAttributes, enabled, baseAccoun
...
 }
 
 /**
- * @desc - Update a subAccount by subAccountID
- * @param subAccountId {string} - The ID generated when using createSubAccount
- * @param [name] {string} - The name displayed in the management console.
- * @param [cloudName] {string} - Unique identifier, used in API calls
- * @param [customAttributes] {object}
- * @param [enabled] {boolean} - Set the sub-account as enabled or not.
+ * @desc - Update a sub_account by sub_account_id
```
Comment (asisayag2) [inline lib/provisioning/account.js]: What remove the "by sub_account_id" part

File: lib/provisioning/account.js [hunk 15]
```diff
@@ -47,118 +57,118 @@ function createSubAccount(name, cloudName, customAttributes, enabled, baseAccoun
...
+    sub_account_ids: sub_account_ids,
   };
   options.content_type = 'json';
   return call_account_api('POST', uri, params, callback, options);
 }
 
 /**
  *
- * @param userId {string} - The userId created by createUser
+ * @param user_id {string} - The user_id created by create_user
```
Comment (asisayag2) [inline lib/provisioning/account.js]: What is `create_user`?

File: lib/network/call_account_api.js (hunk 6 above)
Comment (patrick-tolosa) [reply lib/network/call_account_api.js]: Done

File: lib/config.js (hunk 5 above)
Comment (patrick-tolosa) [reply lib/config.js]: I tried to avoid over-abstraction here, do you really think this refactor is also needed as part of this feature?

File: lib/provisioning/account.js (hunk 14 above)
Comment (patrick-tolosa) [reply lib/provisioning/account.js]: sub_account_id is the identifier by which you perform the action.
Do you have an idea of how to rename it to make it clearer?

(in theory, this is the same syntax as getElementById, getElementByClassName)

File: lib/provisioning/account.js (hunk 8 above)
Comment (patrick-tolosa) [reply lib/provisioning/account.js]: Fixed

File: lib/provisioning/account.js (hunk 9 above)
Comment (patrick-tolosa) [reply lib/provisioning/account.js]: Fixed

File: lib/provisioning/account.js (hunk 10 above)
Comment (patrick-tolosa) [reply lib/provisioning/account.js]: Fixed

File: lib/provisioning/account.js (hunk 11 above)
Comment (patrick-tolosa) [reply lib/provisioning/account.js]: Fixed

File: lib/provisioning/account.js (hunk 13 above)
Comment (asisayag2) [reply lib/provisioning/account.js]: Won't be a blocker for this PR, but let's figure it out and update the docs if needed

File: lib/provisioning/account.js [hunk 16]
```diff
@@ -0,0 +1,292 @@
...
+    email,
+    role,
+    sub_account_ids: sub_account_ids,
+  };
+  options.content_type = 'json';
+  return call_account_api('POST', uri, params, callback, options);
+}
+
+/**
+ * @desc Update a user with values
```
Comment (jackieros) [inline lib/provisioning/account.js]:
```suggestion
 * @desc Update the details of a specified user.
```

File: lib/provisioning/account.js [hunk 17]
```diff
@@ -0,0 +1,292 @@
...
+ */
+function delete_user_group(group_id, options = {}, callback) {
+  let uri = ['user_groups', group_id];
+  return call_account_api('DELETE', uri, {}, callback, options);
+}
+
+/**
+ * @desc Add a user to a group with the specified ID.
+ * @param group_id {string} - The ID of the user group.
+ * @param user_id {string} - The ID of the user
```
Comment (jackieros) [inline lib/provisioning/account.js]:
```suggestion
 * @param user_id {string} - The ID of the user.
```

File: lib/provisioning/account.js [hunk 18]
```diff
@@ -0,0 +1,292 @@
...
+ * @param [options] {object} - @see <a href="https://cloudinary.com/documentation/cloudinary_sdks#configuration_parameters" target="_blank">Documentation</a>
+ * @param [callback] {function}
+ */
+function user_groups(options = {}, callback) {
+  let uri = ['user_groups'];
+  return call_account_api('GET', uri, {}, callback, options);
+}
+
+/**
+ * @desc - Get users in a user group
```
Comment (jackieros) [inline lib/provisioning/account.js]:
```suggestion
 * @desc - Returns the users in the specified group.
```

File: lib/provisioning/account.js [hunk 19]
```diff
@@ -0,0 +1,292 @@
...
+ * @param group_id {string} - The group_id created by account.create_user_group()
```
Comment (jackieros) [inline lib/provisioning/account.js]:
```suggestion
 * @param group_id {string} - The ID of the user group.
```

File: lib/api_client/execute_request.js [hunk 20]
```diff
@@ -0,0 +1,131 @@
...
+        let result;
+        if (error) {
+          return;
+        }
+        try {
+          result = JSON.parse(buffer);
+        } catch (e) {
+          result = {
+            error: {
+              message: "Server return invalid JSON response. Status Code " + res.statusCode,
```
[the remark this answers is not in the document]
Comment (patrick-tolosa) [reply lib/api_client/execute_request.js]: This means that the server returned data not in JSON form, so the SDK can't parse it.

File: lib/provisioning/account.js [hunk 21]
```diff
@@ -0,0 +1,292 @@
...
+function delete_sub_account(sub_account_id, options = {}, callback) {
+  let uri = ['sub_accounts', sub_account_id];
+  return call_account_api('DELETE', uri, {}, callback, options);
+}
+
+/**
+ * @desc - Update the the specified details of the sub-account.
+ * @param sub_account_id {string} - The ID of the sub-account to update.
+ * @param [name] {string} - The display name as shown in the management console.
+ * @param [cloud_name] {string} - A case-insensitive cloud name comprised of alphanumeric and underscore characters. Use with caution - generates an error if the cloud name is not unique across all Cloudinary accounts. Note: Can only be changed for accounts with fewer than 1000 images.
```
[the remark this answers is not in the document]
Comment (patrick-tolosa) [reply lib/provisioning/account.js]: Happy to change, what did you have in mind?

Should we just omit the 'use with caution' part?

File: lib/provisioning/account.js [hunk 22]
```diff
@@ -0,0 +1,308 @@
...
+}
+
+/**
+ * @desc Updates the the specified details of the sub-account.
+ * @param sub_account_id {string} - The ID of the sub-account.
+ * @param [name] {string} - The display name as shown in the management console.
+ * @param [cloud_name] {string} - A case-insensitive cloud name comprised of alphanumeric and underscore characters.
+ *                                Use with caution - generates an error if the cloud name is not unique across all
+ *                                Cloudinary accounts.
+ *                                Note: Can only be changed for accounts with fewer than 1000 images.
```
Comment (jackieros) [inline lib/provisioning/account.js]:
```suggestion
 *                                Notes: 
 *                                - Can only be changed for accounts with fewer than 1000 images.
 *                                - Generates an error if the cloud name is not unique across all Cloudinary accounts. 
```

File: lib/provisioning/account.js [hunk 23]
```diff
@@ -0,0 +1,308 @@
...
+  let uri = ['sub_accounts', sub_account_id];
+  return call_account_api('DELETE', uri, {}, callback, options);
+}
+
+/**
+ * @desc Updates the the specified details of the sub-account.
+ * @param sub_account_id {string} - The ID of the sub-account.
+ * @param [name] {string} - The display name as shown in the management console.
+ * @param [cloud_name] {string} - A case-insensitive cloud name comprised of alphanumeric and underscore characters.
+ *                                Use with caution - generates an error if the cloud name is not unique across all
```
Comment (jackieros) [inline lib/provisioning/account.js]:
```suggestion

```

File: lib/provisioning/account.js [hunk 24]
```diff
@@ -0,0 +1,308 @@
...
+ *                                Cloudinary accounts.
+ *                                Note: Can only be changed for accounts with fewer than 1000 images.
+ * @param [custom_attributes] {object} - Any custom attributes you want to associate with the sub-account, as a map/hash
```
Comment (jackieros) [inline lib/provisioning/account.js]: Is "map/hash" correct for node.js?

Comment (patrick-tolosa) [reply lib/provisioning/account.js]: Yea it's understandable enough, all objects in JS are actually a form of hash.

File: lib/provisioning/account.js [hunk 25]
```diff
@@ -0,0 +1,308 @@
...
+function delete_sub_account(sub_account_id, options = {}, callback) {
+  let uri = ['sub_accounts', sub_account_id];
+  return call_account_api('DELETE', uri, {}, callback, options);
+}
+
+/**
+ * @desc Updates the the specified details of the sub-account.
+ * @param sub_account_id {string} - The ID of the sub-account.
+ * @param [name] {string} - The display name as shown in the management console.
+ * @param [cloud_name] {string} - A case-insensitive cloud name comprised of alphanumeric and underscore characters.
```
[the remark this answers is not in the document]
Comment (patrick-tolosa) [reply lib/provisioning/account.js]: Should these changes be done everywhere the cloud_name field appears? or should we keep the original wording?

File: lib/provisioning/account.js (hunk 23 above)
Comment (const-cloudinary) [reply lib/provisioning/account.js]: This is from @MeirFeinberg , he wanted to emphasise since it returns strange error, but yeah can be removed

File: types/index.d.ts [hunk 26]
```diff
@@ -344,6 +345,9 @@ declare module 'cloudinary' {
         sign_url?: boolean;
         use_root_path?: boolean;
         auth_token?: object;
+        account_id?: string;
+        provisioning_api_key: string;
+        provisioning_api_secret: string;
```
Comment (jifeon) [inline types/index.d.ts:350]: Hi @patrick-tolosa! Why these two are required? My type checking start complaining about them.

Comment (patrick-tolosa) [reply types/index.d.ts:350]: Hi,

Thanks for bringing this up, I've opened #451  to address it.

Comment (jifeon) [reply types/index.d.ts:350]: thank you!
